File: 2-D Dynamic Programming/1277. Count Square Submatrices with All Ones/solution.py
class SolutionTopDown:
    def countSquares(self, matrix: List[List[int]]) -> int:
        m, n = len(matrix), len(matrix[0])
        
        memo = {}
        def dfs(r, c):
            if r == m or c == n: return 0
            
            if (r,c) in memo:
                return memo[(r,c)]
            
            if matrix[r][c] == 1:
                memo[(r,c)] = 1 + min(dfs(r+1, c), dfs(r, c+1), dfs(r+1, c+1))
            else:
                memo[(r,c)] = 0
            return memo[(r,c)]
        
        cnt = 0
        for row in range(m):
            for col in range(n):
                if matrix[row][col]:
                    cnt += dfs(row, col)

        return cnt

class SolutionBottomUp:
    def countSquares(self, matrix: List[List[int]]) -> int:
        m, n = len(matrix), len(matrix[0])
        
        dp = [[0]*(n+1) for _ in range(m+1)]
        for i in range(1, m+1):
            for j in range(1, n+1):
                dp[i][j] = matrix[i-1][j-1]

        total = 0
        for i in range(1, m+1):
            for j in range(1, n+1):
                if matrix[i-1][j-1] == 1:
                    dp[i][j] = 1 + min(dp[i-1][j-1], dp[i-1][j], dp[i][j-1])
                    total +=  dp[i][j]

        # return sum(map(sum, dp))
        return total

class SolutionSpaceOptimized:
    def countSquares(self, matrix: List[List[int]]) -> int:
        m, n = len(matrix), len(matrix[0])
        
        dp, prevDp = [0]*(n+1), [0]*(n+1)

        total = 0
        for i in range(1, m+1):
            for j in range(1, n+1):
                dp[j] = matrix[i-1][j-1] # initialization

                if matrix[i-1][j-1] == 1:
                    dp[j] = 1 + min(prevDp[j-1], prevDp[j], dp[j-1])
                    total += dp[j]
            dp, prevDp = prevDp, dp

        return total

File: 2-D Dynamic Programming/1277. Count Square Submatrices with All Ones/solution.cpp
/*
square: 四邊長相等 => 用邊長長度作為特徵值
dp[i][j]: 在(i,j)位置的合法最長邊長

那麼狀態轉移就考慮左邊目前最長, 上方延伸最長多長, 以及左上角整片的合法最長長度

if matric[i][j] == 1: 更新dp[i][j] = min(dp[i-1][j], dp[i][j-1], dp[i-1][j-1]) + 1

假設dp[i][j] = 2, 代表(i,j)這位置能組出邊長為1跟2兩種正方形 => res += 2
*/

class Solution {
public:
    int countSquares(vector<vector<int>>& matrix) {
        int m = matrix.size(), n = matrix[0].size();

        vector<vector<int>> dp = vector(m, vector<int>(n, 0));
        int res = 0;
        for (int i=0; i<m; i++) {
            for (int j=0; j<n; j++) {
                if (matrix[i][j] == 1) {
                    int top = i-1>=0 ? dp[i-1][j] : 0;
                    int left = j-1>=0 ? dp[i][j-1] : 0;
                    int topleft = i-1>=0 && j-1 >=0 ? dp[i-1][j-1] : 0;
                    dp[i][j] = min({top, left, topleft}) + 1;
                    res += dp[i][j];
                }
            }
        }
        return res;
    }
};
